lib/Common/Memory/RecyclerWriteBarrierManager.cpp

```diff
@@ -71,10 +74,27 @@ X64WriteBarrierCardTableManager::OnThreadInit()
     this->_stacklimit = (char*)stackEnd;
 #endif
 
-    // on Windows server 2012 stack limit can expand with process running, and causes
-    // accessing uncommitted card table page.
-    // TODO: use VirtualQuery twice to get the max possible stack limit
-    stackEnd -= AutoSystemInfo::PageSize * AutoSystemInfo::PageSize;
+#ifdef _WIN32
+    MEMORY_BASIC_INFORMATION memInfo;
+    VirtualQuery((LPCVOID)stackEnd, &memInfo, sizeof(memInfo));
+
+    // using AllocationBase, while allocating stack, the OS reserve a chunk of memory for stack expanding, layout looks like:
+    // ----|-------------------------|-------------------|---------------------------------|---------------------
+    // stackBase  commited     stackLimit GuardPages            reserved region         hard limit
+    // the GuardPages can slide right in the reserved region while stack expanding, until hit the hard limit
+    // the AllocationBase from the VirtualQuery is the whole region possibly used for stack
+    // the BaseAddress from VirtualQuery is current commited base address, we can also query the region of GuardPages but seems not necessary
+
+    stackEnd = (char*)memInfo.AllocationBase;
+    Assert(memInfo.AllocationProtect == PAGE_READWRITE);
+#else
+    pthread_attr_t attr;
+    size_t stacksize;
+    pthread_attr_init(&attr);
+    pthread_attr_getstacksize(&attr, &stacksize); // assuming no one is calling pthread_attr_setstacksize afterwards
```

Comment:
Besides we tend to not do this `ifdef _WIN32` stuff under the core. All the necessary `stack` related functionality is available under PAL or PlatformAgnostic